modules/lotamePanoramaIdSystem.js

```diff
@@ -0,0 +1,243 @@
+}
+
+/**
+ * Clear a cached value from cookies and local storage
+ * @param {String} key
+ */
+function clearLotameCache(key) {
+  if (key) {
+    if (storage.cookiesAreEnabled()) {
+      let expirationDate = new Date(0).toUTCString();
+      storage.setCookie(key, '', expirationDate, 'Lax', undefined, undefined);
+    }
+    if (storage.hasLocalStorage()) {
+      storage.removeDataFromLocalStorage(key, undefined);
+    }
+  }
+}
+/** @type {Submodule} */
+export const lotamePanoramaIdSubmodule = {
+
+  /**
+   * used to link submodule with config
+   * @type {string}
+   */
+  name: MODULE_NAME,
+
+  /**
+   * Decode the stored id value for passing to bid requests
+   * @function decode
+   * @param {(Object|string)} value
+   * @returns {(Object|undefined)}
+   */
+  decode(value, configParams) {
+    return utils.isStr(value) ? { 'lotamePanoramaId': value } : undefined;
+  },
+
+  /**
+   * Retrieve the Lotame Panorama Id
+   * @function
+   * @param {SubmoduleParams} [configParams]
+   * @param {ConsentData} [consentData]
+   * @param {(Object|undefined)} cacheIdObj
+   * @returns {IdResponse|undefined}
+   */
+  getId(configParams, consentData, cacheIdObj) {
+    let localCache = getLotameLocalCache();
+
+    let refreshNeeded = Date.now() > localCache.expiryTimestampMs;
+
+    if (!refreshNeeded) {
+      return {
+        id: localCache.data
+      };
+    }
+
+    const storedUserId = getProfileId();
+
+    const resolveIdFunction = function (callback) {
+      let queryParams = {};
+      if (storedUserId) {
+        queryParams.fp = storedUserId
+      }
+
+      if (consentData && utils.isBoolean(consentData.gdprApplies)) {
+        queryParams.gdpr_applies = consentData.gdprApplies;
+        if (consentData.gdprApplies) {
+          queryParams.gdpr_consent = consentData.consentString;
+        }
+      }
+      const url = utils.buildUrl({
+        protocol: 'https',
+        host: `id.crwdcntrl.net`,
+        pathname: '/id',
+        search: utils.isEmpty(queryParams) ? undefined : queryParams,
+      });
+      ajax(
+        url,
+        (response) => {
+          let responseObj = {};
+          if (response) {
+            try {
+              responseObj = JSON.parse(response);
+              saveLotameCache(KEY_EXPIRY, responseObj.expiry_ts);
+
+              if (utils.isStr(responseObj.core_id)) {
+                saveLotameCache(
+                  KEY_ID,
+                  responseObj.core_id,
+                  responseObj.expiry_ts
+                );
+              } else {
+                clearLotameCache(KEY_ID);
+              }
+
+              if (utils.isStr(responseObj.profile_id)) {
+                setProfileId(responseObj.profile_id);
+              } else {
+                clearLotameCache(KEY_PROFILE);
+                clearLotameCache(KEY_ID);
+              }
+            } catch (error) {
+              utils.logError(error);
+            }
+          }
+          callback(responseObj);
```

Comment:
again, i may be missing something, but here you are storing to the prebid cache object the entire `responseObj`, which seems to contain several keys, including the `core_id` and `profile_id`. however, in your decode function on line 157, you are checking if the `value` is a string, otherwise you return `undefined`. my understanding is that the `value` you receive in `decode()` will be the same thing you pass in the `callback()` here...which means decode would always return undefined.  again, i'm sure i'm missing something here since you're doing some of your own cache management, but wanted to flag this as it doesn't seem correct.

Reply:
Looking into this part as it was tricky to trace what `decode` will always get as there are a couple of different paths

Reply:
Pushed up a fix, thanks for catching this!